src/msql/msqladmin.8
.TH MSQLADMIN 1 "16 February 1994"
.SH NAME
msqladmin   \-  miniSQL database administration
.SH SYNOPSIS
.nf
.ft B
msqladmin  [create | drop] database
.ft
.fi
.SH DESCRIPTION
.LP
msqladmin is used to create and drop entire databases.  It must be run
on the machine acting as the miniSQL server.  It will not work if
executed in a client/server manner.

.SH SEE ALSO
.BR msql (1),
.BR relshow (1),
.BR msql (3),
.BR msqld (8),
\fB"MiniSQL Language Specification"\fP.

.SH AUTHOR
miniSQL has been written by David Hughes of Bond University,
Australia as part of the Minerva Network Management
Environment.  He may be reached via electronic mail using 
bambi@Bond.edu.au


doc/HISTORY
mSQL 1.0 Patch 5        15 - Feb - 1995
---------------------------------------

	o Fixed bug in long text field handling
	o Added C++ support for msql.h
	o Added support for MSQL_HOME environ variable in msqld
	o Added HAVE_MMAP defines around new munmap() call
	o Extracted the install from the Makefile to an external script
	o Replaced "-c -o $obj" with "-c" in the makegen macros
	o Added partial match optimisation for the outer loop of joins
	o Fixed integer handling for Cray (64bit)


mSQL 1.0 Patch 4        8 - Feb - 1995
--------------------------------------

	o Moved target stuff to a script as bash dies on the Makefile
	o Fixed bug in net.c regarding non-terminating reads
	o Fixed memory leak left after 1.0.3 testing
	o Added identifier length checks
	o Fixed Multiple free problem with tmp tables values
	o Fixed memory leak with tmp condition list values
	o Fixed formatPacket() to handle long char fields (>1024)
	o Added hash based keyword lookup to parser  (speedup)
	o Reduced the use of bzero and bcopy (speedup)
	o Inlined a couple of functions that are called frequently (speedup)


mSQL 1.0 Patch 3        1 - Feb - 1995
--------------------------------------

	o Fixed conInfo index bug
	o Fixed net.c to ensure full reads and writes
	o Added 64-bit int support for Crays
	o Fixed Cray stuff in the README
	o Added msqlClose() to msql.h
	o Added AIX sys/select.h support to common/config.h.in
	o Added further version info to msqladmin
	o Added calls to msqlClose() in relshow
	o Added error reporting for multiple primary keys in create
	o Fixed key handling in "complex" conditions
	o Added socket details to debugging output (general in msqld
		and api in libmsql).
	o Started on primary key support withing joins
	o Reverted to an older version of autoconf so that it
		will build under SunOS 4.1.3

mSQL 1.0 Patch 2        18 - Jan - 1995
--------------------------------------

	o Fixed mmap() problem with key remapping
	o fixed -h option on msql
	o fixed tabs in # defines in site.h for AIX
	o Added "grep -i" to killer test for AIX's time output
	o Made net.c a bit more robust
	o Fixed LIKE field over-run problem
	o Added AIX stuff to sys-arch
	o Upgraded to the current version of autoconf
	o Fixed long field/table name over-runs (core dump)
	o Tightened up file and directory modes
	o Added test for max number of connections
	o Fixed socket handling on failed connection in libmsql
	o Fixed makegen macros for SVR4.2


mSQL 1.0 Patch 1        5 - Jan - 1995
--------------------------------------

	o Added -h option to msql
	o Added MSQL_HOST support to msql
	o Fixed arg count in call to msqlLoadAcl() in msqld.c
	o Fixed fchmod() problem in makedepend for SCO
	o Added ACL support for admin commands from "localhost" as
		SCO doesn't have UNIX sock's so admin commands come
		from localhost over TCP.
	o Fixed NULL value insert and condition core dumps
	o Fixed comments within char literals in the terminal monitor
	o Added -q mode for msqladmin 
	o Changed HOST to MSQLHOST in rtest to overcome default environments
	o Removed myself from run_daemon so that I don't get other
		peoples crash reports
	o Added mode setting for install directories to mkinstalldirs
	o Fixed install.mm and libinstall.mm to only chown if owned by root
	o Fixed yyerror() so that a NULL yytext doesn't cause a SEGV
	o Added quote escaping to msqldump
	o Added "integer" and "smallint" to the lexer
	o Fixed field list over-run and protocol screwup in msqlListFields()
	o Added on-the-fly INT to REAL conversion on inserts
	o Added filename and line number info to malloc and mmap debugging
	o Fixed huge memory leak in ident structure handling
	o Fixed remapping during keyed inserts (50% less map/unmap calls)
	o Fixed int active comparisons
	o Forced a buffer reload after deleteRow() if mmap() not in use.
		Thanks to Pierre Dupre (Pierre.Dupre@metro.fr) for
		finding and fixing this problem on his Motorola.
	o Fixed NULL row free in createSortedTable(), createDistinctTable()
	o Added malloc() and mmap() leak detection in debug modes
	o Added table cache cleanup on exit.
	o Added continuation prompting to monitor



mSQL 1.0                14 - Dec - 1994
---------------------------------------

	o Fixed select loop bug (bad FD on partial client connection)
	o Added read loop timeout for those with interruptable syscalls
	o Fixed one-line bug in readKey() that cause key lookups to fail
	o Merged msqlsave by Igor Romanenko (igor@frog.kiev.ua) into the
		distribution as msqldump.  Added ability to dump entire
		databases and reformated to match the rest of the code.
	o Fixed call to /bin/echo in makegen for bash
	o Added #ifdef's for signal names around calls to signal()
	o Added DISTINCT operator to SELECT's
	o Fixed bad type checking in writeKey()
	o Fixed key remapping bug (wasn't remapped between inserts).
	o Ported code to FreeBSD 2.0 (including getting around a
		compiler bug and a _very_ wierd shell!)
	o Recoded the lex scanner in C (msql_lex.c)
	o Added support for escaped data in text literals
	o Recoded text literal code to handle 8-bit data rather than
		just strings (i.e. no strlen(), strcpy() etc.)
	o Added Version info to the server
	o Added 'version' command to msqladmin
	o Added '-h host' option to all commands
	o Removed support for MSQL_HOST env var as -h does the job better
	o Added '-q' flag to msql 
	o Added regession test suite


mSQL 0.3 Beta Patch 1	28 - Nov - 1994
---------------------------------------

	o Fixed bug that caused the server to go into an endless read loop
		on a bad client crash (SIGPIPE related)
	o Added argc test to msqladmin
	o Fixed POSIX_SOURCE garbage in X's headers for makedepend
	o Recoded signal handling in makedepend/main.c so as not to
		use sigaction() and friends (just use signal() )
	o Added "proctitle" debugging option
	o Added "mmap" debugging option
	o Fixed mmap() key file seg fault after re-map
	o Generally cleanup up the mmap() based code.

src/msql/msqladmin.c
/*
**	msqladmin.c	- 
**
**
** Copyright (c) 1993  David J. Hughes
**
** Permission to use, copy, and distribute for non-commercial purposes,
** is hereby granted without fee, providing that the above copyright
** notice appear in all copies and that both the copyright notice and this
** permission notice appear in supporting documentation.
**
** This software is provided "as is" without any expressed or implied warranty.
**
** ID = "$Id:"
**
*/


#include <stdio.h>
#include <sys/types.h>
#include <fcntl.h>

#ifdef HAVE_DIRENT
#  include <dirent.h>
#else
#  include <sys/dir.h>
#endif

#include "version.h"
#include "msql.h"


#include <common/site.h>
#include <common/portability.h>

int	qFlag = 0;
char	*msqlHomeDir;


usage()
{
	printf("\n\nusage : msqladmin [-h host] [-q] <Command>\n\n");
	printf("where command =");
	printf("\t drop DatabaseName\n");
	printf("\t\t create DatabaseName\n");
	printf("\t\t shutdown\n");
	printf("\t\t reload\n");
	printf("\t\t version\n");
	printf("\n -q\tQuiet mode.  No verification of commands.\n\n");
}


createDB(sock,db)
	int	sock;
	char	*db;
{
	if(msqlCreateDB(sock,db) < 0)
	{
		fprintf(stderr,"\nmSQL Command failed!\nServer error = %s\n\n",
				msqlErrMsg);
		msqlClose(sock);
		exit(1);
	}
	else
	{
		printf("Database \"%s\" created.\n",db);
	}
}


dropDB(sock,db)
	int	sock;
	char	*db;
{
	char	buf[10];


	if (!qFlag)
	{
		printf("\n\nDropping the database is potentially a very bad ");
		printf("thing to do.\nAny data stored in the database will be");
		printf(" destroyed.\n\nDo you really want to drop the ");
		printf("\"%s\" ",db);
		printf("database?  [Y/N] ");
		bzero(buf,10);
		fgets(buf,10,stdin);
		if ( (*buf != 'y') && (*buf != 'Y'))
		{
			printf("\n\nOK, aborting database drop!\n\n");
			msqlClose(sock);
			exit(0);
		}
	}
	if(msqlDropDB(sock,db) < 0)
	{
		fprintf(stderr,"\nmSQL Command failed!\nServer error = %s\n\n",
			msqlErrMsg);
		msqlClose(sock);
		exit(1);
	}
	else
	{
		fprintf(stderr,"Database \"%s\" dropped\n",db);
	}
}



main(argc,argv)
	int	argc;
	char	*argv[];
{
	int	sock,
		c,
		argsLeft,
		errFlag = 0;
	char	*host = NULL;
	extern	int optind;
	extern	char *optarg;


        msqlHomeDir = (char *)getenv("MSQL_HOME");
        if (!msqlHomeDir)
        {
                msqlHomeDir = INST_DIR;
        }

	while((c=getopt(argc,argv,"h:q"))!= -1)
        {
                switch(c)
                {
                        case 'h':
                                if (host)
                                        errFlag++;
                                else
                                        host = optarg;
                                break;
			case 'q':
				if (qFlag)
					errFlag++;
				else
					qFlag++;
				break;
			case '?':
				errFlag++;
				break;
		}
	}

	argsLeft = argc - optind;

	if (errFlag || argsLeft == 0)
	{
		usage();
		exit(1);
	}


        if ((sock = msqlConnect(host)) < 0)
        {
                fprintf(stderr,"ERROR : %s\n",msqlErrMsg);
                exit(1);
        }

	if (strcmp(argv[optind],"create") == 0)
	{
		if (argsLeft != 2)
		{
			usage();
			msqlClose(sock);
			exit(1);
		}
		createDB(sock,argv[optind+1]);
		msqlClose(sock);
		exit(0);
	}
	if (strcmp(argv[optind],"drop") == 0)
	{
		if (argsLeft != 2)
		{
			usage();
			msqlClose(sock);
			exit(1);
		}
		dropDB(sock,argv[optind+1]);
		msqlClose(sock);
		exit(0);
	}
	if (strcmp(argv[optind],"shutdown") == 0)
	{
		if (argsLeft != 1)
		{
			usage();
			msqlClose(sock);
			exit(1);
		}
		if(msqlShutdown(sock) < 0)
		{
			printf("\nmSQL Command failed!\nServer error = %s\n\n",
				msqlErrMsg);
			msqlClose(sock);
			exit(1);
		}
		exit(0);
	}
	if (strcmp(argv[optind],"reload") == 0)
	{
		if (argsLeft != 1)
		{
			usage();
			msqlClose(sock);
			exit(1);
		}
		if(msqlReloadAcls(sock) < 0)
		{
			printf("\nmSQL Command failed!\nServer error = %s\n\n",
				msqlErrMsg);
			msqlClose(sock);
			exit(1);
		}
		msqlClose(sock);
		exit(0);
	}
	if (strcmp(argv[optind],"version") == 0)
	{
		if (argsLeft != 1)
		{
			usage();
			msqlClose(sock);
			exit(1);
		}
		printf("\nVersion Details :-\n\n");
		printf("\tmsqladmin version \t%s\n",SERVER_VERSION);
		printf("\tmSQL connection \t%s\n",msqlGetHostInfo());
		printf("\tmSQL server version \t%s\n", msqlGetServerInfo());
		printf("\tmSQL protocol version \t%d\n", msqlGetProtoInfo());
		printf("\tmSQL TCP socket \t%d\n", MSQL_PORT);
		printf("\tmSQL UNIX socket \t%s\n", MSQL_UNIX_ADDR);
		printf("\tmSQL root user \t\t%s\n", ROOT);
		msqlClose(sock);
		exit(0);
	}
	usage();
	msqlClose(sock);
	exit(1);
}
